Automattic/sensei
# Add/course theme contact teacher

Fixes #4376 

### Changes proposed in this Pull Request

* Adds support for sensei-modal. See more details at `assets/js/sensei-modal.js`.
  * Has very basic features and needs further enhancement, especially in terms of accessibility. Currently the focus is not transferred to the modal content when modal opens. And it's not transferred back to the modal opener button when it is closed.
* Introduces messages rest api.
* Contact-Teacher block now has different behavior and different styling depending on where it is rendered.
  * Posts messages as a regular form when rendered as part of the post content.
  * Sends messages via ajax and shows success message if rendered inside the course theme sidebar.
  * Opens the contact teacher form modal with js.
  * Opens the contact teacher form modal without js.
  
### Issues

* There is a small issue with multiple contact-teacher blocks without js enabled. When clicked on the button the url is updated to indicate to the block that the modal should be opened. But if there are multiple contact teacher blocks on the same page then multiple modals are open at the same time and it looks not ideal. See how it looks in the last video below.

### Testing instructions


* Confirm that contact-teacher button works properly and it is possible to send messages to teacher.
* Disable JavaScript and confirm it is still possible to use contact-teacher form and send messages.
* Enable JavaScript back again and enable the feature flag: `add_filter( 'sensei_feature_flag_course_theme', '__return_true' );`
* Confirm that contact-teacher button on the course theme sidebar works and can send messages.
* Confirm that contact-teacher button on the course theme lesson content works and can send messages.
* Disable JavaScript and repeat two steps above. 


### New/Updated Hooks

* No new hooks but new rest-api is exposed. See `includes/rest-api/class-sensei-rest-api-messages-routes-controller.php`


### Screenshot / Video

#### Contact Teacher block

https://user-images.githubusercontent.com/2578542/145012162-adbef26f-658d-44dc-87a1-44f2aa226bcb.mp4

#### Contact Teacher block without js

https://user-images.githubusercontent.com/2578542/145012319-1de887f8-cc7f-413f-81d3-265fb9bbf1f5.mp4

#### Contact Teacher block (Course Theme enabled)

https://user-images.githubusercontent.com/2578542/145012444-3e787d6d-64d2-415b-878a-618073955b7f.mp4

#### Contact Teacher block without js (Course Theme enabled)

https://user-images.githubusercontent.com/2578542/145012523-5fc5f0e2-7481-468f-ad17-38266f85bf44.mp4
